src/task_change.rs

```diff
@@ -382,49 +396,64 @@ pub fn compute_changeset(
-            .iter_mut()
-            .min_by(|left, right| cmp_tasks_3way(&x, &left.orig, &right.orig))
-            .and_then(|t| {
-                if is_task_admissible(&x, &t.orig, allowed_divergence) {
-                    Some(t)
-                } else {
-                    None
-                }
-            });
-        if let Some(best) = best_match {
-            best.to.push(x);
-        } else {
-            new_tasks.push(x);
-        }
-    }
-
-    // Retrieve changes
-    let changes = changeset
-        .into_iter()
-        .map(|t| {
-            let changes = t
-                .to
-                .iter()
-                .enumerate()
-                .map(|(i, to)| changes_between(&t.orig, &to, i == 0))
-                .collect();
-            (t.orig, changes)
+    // Compute for each task the candidate matches, ordered by preference
+    let from_preferences_matrix: Vec<Vec<usize>> = from.iter()
+        .map(|t| preferred_task_ids(&t, &to, allowed_divergence))
+        .collect();
+    let to_preferences_matrix: Vec<Vec<usize>> = to.iter()
+        .map(|t| preferred_task_ids(&t, &from, allowed_divergence))
+        .collect();
+
+    // Compute a stable matching between the two task lists
+    let (matching, _) =
+            stable_marriage::stable_marriage(from_preferences_matrix, to_preferences_matrix);
+
+    let mut to: Vec<Option<Task>> = to.into_iter()
+        .map(Some)
+        .collect();
+    let mut from: Vec<Option<Task>> = from.into_iter()
+        .map(Some)
+        .collect();
+
+    // Extract changed, new, and deleted tasks
+    let matches: Vec<(Task, Task)> = matching.into_iter()
+        .enumerate()
+        .filter_map(|(i, x)| x.map(|x| (i,x)))
+        .map(|(i, j)| (from[i].take().unwrap(), to[j].take().unwrap()))
+        .collect();
+
+    let new_tasks = to.into_iter()
+        .flat_map(|x| x);
+
+    let deleted_tasks = from.into_iter()
+        .flat_map(|x| x);
+
+    let mut changes = matches.into_iter()
+        .map(|(l, r)| {
+            let changes = changes_between(&l, &r, true);
+            (l, vec![changes])
         })
+        .chain(deleted_tasks.map(|t| (t, vec![])))
         .collect::<Vec<(Task, Vec<Vec<Changes>>)>>();
 
+    // Detect recurred tasks
+    let new_tasks: Vec<Task> = new_tasks
+        .filter(|x| {
+            let best_match = changes
+                .iter_mut()
+                .filter(|(t, _)| t.tags.get("rec").is_some())
+                .filter(|(_, delta)| delta.len() != 0)
+                .filter(|(t, _)| is_task_admissible(&x, t, allowed_divergence))
+                .min_by(|(left, _), (right, _)| cmp_tasks_3way(&x, &left, &right));
+            if let Some((ref mut best, ref mut delta)) = best_match {
+                let changes = changes_between(&best, &x, false);
+                delta.push(changes);
+                false
+            } else {
+                true
+            }
+        }).collect();
```

Comment:
The codebase until now uses the `let x = foo.collect::<Vec<_>>()` idiom instead of the `let x: Vec<_> = foo.collect()` one, so I guess it'd be better to adapt it here.